orchagent/pfcwdorch.cpp

```diff
@@ -320,15 +302,61 @@ void PfcWdOrch<DropHandler, ForwardHandler>::unregisterFromWdDb(const Port& port
         sai_object_id_t queueId = port.m_queue_ids[i];
 
         // Unregister in syncd
-        PfcWdOrch<DropHandler, ForwardHandler>::getPfcWdTable()->del(sai_serialize_object_id(queueId));
+        m_pfcWdTable->del(sai_serialize_object_id(queueId));
+        m_entryMap.erase(queueId);
     }
 }
 
 template <typename DropHandler, typename ForwardHandler>
-PfcWdSwOrch<DropHandler, ForwardHandler>::PfcWdSwOrch(DBConnector *db, vector<string> &tableNames):
-    PfcWdOrch<DropHandler, ForwardHandler>(db, tableNames)
+PfcWdSwOrch<DropHandler, ForwardHandler>::PfcWdSwOrch(
+        DBConnector *db,
+        vector<string> &tableNames,
+        vector<sai_port_stat_t> portStatIds,
+        vector<sai_queue_stat_t> queueStatIds):
+    PfcWdOrch<DropHandler,
+    ForwardHandler>(db, tableNames),
+    m_pfcWdDb(new DBConnector(PFC_WD_DB, DBConnector::DEFAULT_UNIXSOCKET, 0)),
+    m_pfcWdTable(new ProducerStateTable(m_pfcWdDb.get(), PFC_WD_STATE_TABLE)),
+    c_portStatIds(portStatIds),
+    c_queueStatIds(queueStatIds)
 {
     SWSS_LOG_ENTER();
+
+    string platform = getenv("platform") ? getenv("platform") : "";
+
+    if (platform == "")
+    {
+        SWSS_LOG_ERROR("Platform environment variable is not defined");
+        return;
+    }
+
+    string detectSha, restoreSha;
+    string detectPluginName = "pfc_detect_" + platform + ".lua";
+    string restorePluginName = "pfc_restore_" + platform + ".lua";
+
+    try
+    {
+        string detectLuaScript = swss::loadLuaScript(detectPluginName);
+        detectSha = swss::loadRedisScript(
+                PfcWdOrch<DropHandler, ForwardHandler>::getCountersDb().get(),
+                detectLuaScript);
+
+        string restoreLuaScript = swss::loadLuaScript(restorePluginName);
+        restoreSha = swss::loadRedisScript(
+                PfcWdOrch<DropHandler, ForwardHandler>::getCountersDb().get(),
+                restoreLuaScript);
+
+        vector<FieldValueTuple> fieldValues;
+        fieldValues.emplace_back(SAI_OBJECT_TYPE, sai_serialize_object_type(SAI_OBJECT_TYPE_QUEUE));
```

Comment:
why SAI_OBJECT_TYPE_QUEUE? why this plugin table is related to SAI_OBJECT_TYPE_QUEUE?

Reply:
Client registers to a specific type of objects. It will receive a list of queue oids for which thread updated counters.

Reply:
see my comments on sonic-sairedis repo, this does not have to reuse SAI OBJECT TYPE, it is plugin table, you can define your own attribute and value.

Reply:
I will change in a follow-up PR, since it requires change in swss-common schema